### PR TITLE
[Fix](auto-partition) fix auto partition load lost data in multi sender

### PR DESCRIPTION
**Pull request by zclllyybb**
## Proposed changes

Issue Number: close #xxx

Before if we have multi sender but some of their sink node have no data of some specific partition, then this sink won't know it has to close it. So tablets channel won't be closed correctly.
Now change close time needed count to same with open in tablets channel.

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

File: be/src/vec/sink/writer/vtablet_writer.cpp [hunk 1]
```diff
@@ -969,17 +980,50 @@ void VNodeChannel::_close_check() {
     CHECK(_cur_mutable_block == nullptr) << name();
 }
 
+void VNodeChannel::close_first_stage() {
+    auto st = none_of({_cancelled, _eos_is_produced});
+    if (!st.ok()) {
+        return;
```
Comment (HappenLee) [inline be/src/vec/sink/writer/vtablet_writer.cpp:986]: Add warning log

File: be/src/vec/sink/writer/vtablet_writer.cpp [hunk 2]
```diff
@@ -359,22 +357,23 @@ Status VNodeChannel::init(RuntimeState* state) {
...
     _send_block_callback = WriteBlockCallback<PTabletWriterAddBlockResult>::create_shared();
-    _send_block_callback->addFailedHandler([&, task_exec_ctx = _task_exec_ctx](bool is_last_rpc) {
-        auto ctx_lock = task_exec_ctx.lock();
-        if (ctx_lock == nullptr) {
-            return;
-        }
-        _add_block_failed_callback(is_last_rpc);
-    });
+    _send_block_callback->addFailedHandler(
+            [&, task_exec_ctx = _task_exec_ctx](bool is_last_rpc, bool first_stage_close) {
```
Comment (HappenLee) [inline be/src/vec/sink/writer/vtablet_writer.cpp:361]: move `first_stage_close` to exec_ctx.

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

File: be/src/runtime/load_channel.cpp [hunk 3]
```diff
@@ -195,6 +196,22 @@ Status LoadChannel::_handle_eos(BaseTabletsChannel* channel,
     auto index_id = request.index_id();
 
     RETURN_IF_ERROR(channel->close(this, request, response, &finished));
+
+    // for init node, we close waiting(hang on) all close request and let them return together.
+    if (!channel->is_incremental_channel()) {
```
Comment (zclllyybb) [inline be/src/runtime/load_channel.cpp:201]: check auto partition

File: be/src/runtime/tablets_channel.cpp [hunk 4]
```diff
@@ -142,9 +142,16 @@ Status BaseTabletsChannel::open(const PTabletWriterOpenRequest& request) {
     RETURN_IF_ERROR(_schema->init(request.schema()));
     _tuple_desc = _schema->tuple_desc();
 
-    _num_remaining_senders = request.num_senders();
-    _next_seqs.resize(_num_remaining_senders, 0);
-    _closed_senders.Reset(_num_remaining_senders);
+    int max_sender = request.num_senders();
+    // this tablets channel created by inc_open: open times = how many sender instance know this tablets channel.
```
Comment (zclllyybb) [inline be/src/runtime/tablets_channel.cpp:146]: add comment

**Comment by zclllyybb** [comment]
run buildall

File: be/src/vec/sink/writer/vtablet_writer.h [hunk 5]
```diff
@@ -114,6 +106,10 @@ struct AddBatchCounter {
     }
 };
 
+struct WriteBlockCallbackContext {
```
Comment (liaoxin01) [inline be/src/vec/sink/writer/vtablet_writer.h:109]: Only one variable, no need this structure？

File: be/src/vec/sink/writer/vtablet_writer.cpp [hunk 6]
```diff
@@ -1411,7 +1414,6 @@ Status VTabletWriter::_send_new_partition_batch() {
 }
 
 void VTabletWriter::_do_try_close(RuntimeState* state, const Status& exec_status) {
-    SCOPED_TIMER(_close_timer);
```
[the remark this answers is not in the document]
Comment (zclllyybb) [reply be/src/vec/sink/writer/vtablet_writer.cpp:1370]: fixed. thx~

File: be/src/vec/sink/writer/vtablet_writer.h (hunk 5 above)
Comment (zclllyybb) [reply be/src/vec/sink/writer/vtablet_writer.h:109]: introduced in progress of this pr. for more change possible in the future I think this is more friendly so didn't remove this.

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
close this since https://github.com/apache/doris/pull/35287 covered

**Pull request closed without merging**